File: firefox/src/WebCLDevice.cpp
/*
 * This file is part of WebCL – Web Computing Language.
 *
 * This Source Code Form is subject to the terms of the
 * Mozilla Public License, v. 2.0. If a copy of the MPL
 * was not distributed with this file, You can obtain
 * one at http://mozilla.org/MPL/2.0/.
 *
 * The Original Contributor of this Source Code Form is
 * Nokia Research Center Tampere (http://webcl.nokiaresearch.com).
 *
 */

#include "WebCLDevice.h"
#include "WebCLCommon.h"
#include "WebCL_internal.h"
#include "WebCL_clwrapper.h"

#include "nsComponentManagerUtils.h"
#include "nsISecurityCheckedComponent.h"


NS_IMPL_ISUPPORTS2 (WebCLDevice, IWebCLDevice, nsISecurityCheckedComponent)
WEBCL_SECURITY_CHECKED_IMPL (WebCLDevice)
WEBCL_ATTACHMENT_IMPL (WebCLDevice)


/* static */
InstanceRegistry<cl_device_id, WebCLDevice*> WebCLDevice::instanceRegistry;


/* static */
nsresult WebCLDevice::getInstance (cl_device_id aInternal, WebCLDevice** aResultOut,
                                   WebCL_LibCLWrapper* aLibWrapper)
{
  nsresult rv = NS_OK;

  WebCLDevice* existing = 0;
  if (instanceRegistry.findById (aInternal, &existing))
  {
    NS_IF_ADDREF (*aResultOut = existing);
  }
  else
  {
    nsCOMPtr<WebCLDevice> obj ( new WebCLDevice () );
    if (!obj)
    {
      D_LOG (LOG_LEVEL_ERROR, "Failed to create instance. rv=%d.", rv);
      return NS_ERROR_OUT_OF_MEMORY;
    }

    obj->setWrapper (aLibWrapper);
    obj->mInternal = aInternal;

    instanceRegistry.add (obj->mInternal, obj);

    NS_IF_ADDREF (*aResultOut = obj);
  }

  return rv;
}


WebCLDevice::WebCLDevice()
  : IWebCLDevice(), WebCLCommon(),
    mInternal(0)
{
  D_METHOD_START;
}


WebCLDevice::~WebCLDevice()
{
  D_METHOD_START;
  if (mInternal)
    instanceRegistry.remove (mInternal);
}


int WebCLDevice::getTypeForInfoName (int aName)
{
  D_METHOD_START;
  switch (aName)
  {
    case CL_DEVICE_ADDRESS_BITS: return types::UINT;
    case CL_DEVICE_AVAILABLE: return types::BOOL;
    case CL_DEVICE_COMPILER_AVAILABLE: return types::BOOL;
    //case CL_DEVICE_DOUBLE_FP_CONFIG: return types::DEVICE_FP_CONFIG;
    case CL_DEVICE_ENDIAN_LITTLE: return types::BOOL;
    case CL_DEVICE_ERROR_CORRECTION_SUPPORT: return types::BOOL;
    case CL_DEVICE_EXECUTION_CAPABILITIES: return types::DEVICE_EXEC_CAPABILITIES;
    case CL_DEVICE_EXTENSIONS: return types::STRING;
    case CL_DEVICE_GLOBAL_MEM_CACHE_SIZE: return types::ULONG;
    case CL_DEVICE_GLOBAL_MEM_CACHE_TYPE: return types::DEVICE_MEM_CACHE_TYPE;
    case CL_DEVICE_GLOBAL_MEM_CACHELINE_SIZE: return types::UINT;
    case CL_DEVICE_GLOBAL_MEM_SIZE: return types::ULONG;
    //case CL_DEVICE_HALF_FP_CONFIG: return types::DEVICE_HALF_FP_CONFIG;
    case CL_DEVICE_IMAGE_SUPPORT: return types::BOOL;
    case CL_DEVICE_IMAGE2D_MAX_HEIGHT: return types::SIZE_T;
    case CL_DEVICE_IMAGE2D_MAX_WIDTH: return types::SIZE_T;
    case CL_DEVICE_IMAGE3D_MAX_DEPTH: return types::SIZE_T;
    case CL_DEVICE_IMAGE3D_MAX_HEIGHT: return types::SIZE_T;
    case CL_DEVICE_IMAGE3D_MAX_WIDTH: return types::SIZE_T;
    case CL_DEVICE_LOCAL_MEM_SIZE: return types::ULONG;
    case CL_DEVICE_LOCAL_MEM_TYPE: return types::DEVICE_LOCAL_MEM_TYPE;
    case CL_DEVICE_MAX_CLOCK_FREQUENCY: return types::UINT;
    case CL_DEVICE_MAX_COMPUTE_UNITS: return types::UINT;
    case CL_DEVICE_MAX_CONSTANT_ARGS: return types::UINT;
    case CL_DEVICE_MAX_CONSTANT_BUFFER_SIZE: return types::ULONG;
    case CL_DEVICE_MAX_MEM_ALLOC_SIZE: return types::ULONG;
    case CL_DEVICE_MAX_PARAMETER_SIZE: return types::SIZE_T;
    case CL_DEVICE_MAX_READ_IMAGE_ARGS: return types::UINT;
    case CL_DEVICE_MAX_SAMPLERS: return types::UINT;
    case CL_DEVICE_MAX_WORK_GROUP_SIZE: return types::SIZE_T;
    case CL_DEVICE_MAX_WORK_ITEM_DIMENSIONS: return types::UINT;
    case CL_DEVICE_MAX_WORK_ITEM_SIZES: return types::SIZE_T_V;
    case CL_DEVICE_MAX_WRITE_IMAGE_ARGS: return types::UINT;
    case CL_DEVICE_MEM_BASE_ADDR_ALIGN: return types::UINT;
    case CL_DEVICE_MIN_DATA_TYPE_ALIGN_SIZE: return types::UINT;
    case CL_DEVICE_NAME: return types::STRING;
    case CL_DEVICE_PLATFORM: return types::PLATFORM;
    case CL_DEVICE_PREFERRED_VECTOR_WIDTH_CHAR: return types::UINT;
    case CL_DEVICE_PREFERRED_VECTOR_WIDTH_SHORT: return types::UINT;
    case CL_DEVICE_PREFERRED_VECTOR_WIDTH_INT: return types::UINT;
    case CL_DEVICE_PREFERRED_VECTOR_WIDTH_LONG: return types::UINT;
    case CL_DEVICE_PREFERRED_VECTOR_WIDTH_FLOAT: return types::UINT;
    case CL_DEVICE_PREFERRED_VECTOR_WIDTH_DOUBLE: return types::UINT;
    case CL_DEVICE_PROFILE: return types::STRING;
    case CL_DEVICE_PROFILING_TIMER_RESOLUTION: return types::SIZE_T;
    case CL_DEVICE_QUEUE_PROPERTIES: return types::COMMAND_QUEUE_PROPERTIES;
    case CL_DEVICE_SINGLE_FP_CONFIG: return types::DEVICE_FP_CONFIG;
    case CL_DEVICE_TYPE: return types::DEVICE_TYPE;
    case CL_DEVICE_VENDOR: return types::STRING;
    case CL_DEVICE_VENDOR_ID: return types::UINT;
    case CL_DEVICE_VERSION: return types::STRING;
    case CL_DRIVER_VERSION: return types::STRING;
    default: ;
  }
  return types::UNKNOWN;
}


/* nsIVariant getDeviceInfo (in long aName); */
NS_IMETHODIMP WebCLDevice::GetDeviceInfo(PRInt32 aName, JSContext *cx, nsIVariant **_retval)
{
  D_METHOD_START;

  NS_ENSURE_ARG_POINTER (_retval);
  nsresult rv;
  cl_int err = CL_SUCCESS;
  int type = getTypeForInfoName (aName);
  if (type == types::UNKNOWN)
  {
    D_LOG (LOG_LEVEL_ERROR, "Info parameter name %d does not have a known type.", aName);
    WebCL_reportJSError (cx, "Info name %d is not supported by %s.",
                         aName, __FUNCTION__);
    return WEBCL_XPCOM_ERROR; //NS_ERROR_FAILURE;
  }

  nsCOMPtr<nsIWritableVariant> variant = do_CreateInstance(NS_VARIANT_CONTRACTID, &rv);
  NS_ENSURE_SUCCESS (rv, rv);

  WEBCL_GETINFO_MEDIATOR_SWITCH (aName, type, mWrapper, getDeviceInfo, mInternal,
                                 variant, err, rv);

  NS_ADDREF (*_retval = variant);

  return NS_OK;
}

File: firefox-js/js/webcldevice.js
/*
 * This file is part of WebCL – Web Computing Language.
 *
 * This Source Code Form is subject to the terms of the
 * Mozilla Public License, v. 2.0. If a copy of the MPL
 * was not distributed with this file, You can obtain
 * one at http://mozilla.org/MPL/2.0/.
 *
 * The Original Contributor of this Source Code Form is
 * Nokia Research Center Tampere (http://webcl.nokiaresearch.com).
 *
 */


const Cc = Components.classes;
const Ci = Components.interfaces;
const Cu = Components.utils;

var Exception = Components.Exception;


Cu.import ("resource://gre/modules/Services.jsm");
Cu.import ("resource://gre/modules/XPCOMUtils.jsm");
Cu.import ("resource://nrcwebcl/modules/logger.jsm");
Cu.import ("resource://nrcwebcl/modules/webclutils.jsm");
Cu.import ("resource://nrcwebcl/modules/base.jsm");

Cu.import ("resource://nrcwebcl/modules/lib_ocl/ocl_constants.jsm");
Cu.import ("resource://nrcwebcl/modules/lib_ocl/ocl_exception.jsm");


try {


var CLASSNAME =  "WebCLDevice";
var CID =        "{f5352722-9a35-405b-95ae-54d5b4995576}";
var CONTRACTID = "@webcl.nokiaresearch.com/IWebCLDevice;1";


function Device ()
{
  if (!(this instanceof Device)) return new Device ();

  Base.apply(this);

  this.wrappedJSObject = this;

  this._interfaces = [ Ci.IWebCLDevice,
                       Ci.nsISecurityCheckedComponent,
                       Ci.nsISupportsWeakReference,
                       Ci.nsIClassInfo,
                       Ci.nsISupports
                     ];
}

Device.prototype = Object.create (Base.prototype);

Device.prototype.classDescription = CLASSNAME;
Device.prototype.classID =          Components.ID(CID);
Device.prototype.contractID =       CONTRACTID;
Device.prototype.QueryInterface =   XPCOMUtils.generateQI ([ Ci.IWebCLDevice,
                                                             Ci.nsISecurityCheckedComponent,
                                                             Ci.nsISupportsWeakReference,
                                                             Ci.nsIClassInfo
                                                           ]);


//------------------------------------------------------------------------------
// IWebCLDevice


Device.prototype.getInfo = function (name)
{
  TRACE (this, "getInfo", arguments);
  if(!this._ensureValidObject ()) throw new CLInvalidated();

  try
  {
    switch (name)
    {
      case ocl_info.CL_DEVICE_TYPE:                               break;
      case ocl_info.CL_DEVICE_VENDOR_ID:                          break;
      case ocl_info.CL_DEVICE_MAX_COMPUTE_UNITS:                  break;
      case ocl_info.CL_DEVICE_MAX_WORK_ITEM_DIMENSIONS:           break;
      case ocl_info.CL_DEVICE_MAX_WORK_ITEM_SIZES:                break;
      case ocl_info.CL_DEVICE_MAX_WORK_GROUP_SIZE:                break;
      case ocl_info.CL_DEVICE_PREFERRED_VECTOR_WIDTH_CHAR:        break;
      case ocl_info.CL_DEVICE_PREFERRED_VECTOR_WIDTH_SHORT:       break;
      case ocl_info.CL_DEVICE_PREFERRED_VECTOR_WIDTH_INT:         break;
      case ocl_info.CL_DEVICE_PREFERRED_VECTOR_WIDTH_LONG:        break;
      case ocl_info.CL_DEVICE_PREFERRED_VECTOR_WIDTH_FLOAT:       break;
      case ocl_info.CL_DEVICE_PREFERRED_VECTOR_WIDTH_DOUBLE:      throw new CLError(ocl_errors.CL_INVALID_VALUE);
      case ocl_info.CL_DEVICE_PREFERRED_VECTOR_WIDTH_HALF:        throw new CLError(ocl_errors.CL_INVALID_VALUE);
      case ocl_info.CL_DEVICE_NATIVE_VECTOR_WIDTH_CHAR:           break;
      case ocl_info.CL_DEVICE_NATIVE_VECTOR_WIDTH_SHORT:          break;
      case ocl_info.CL_DEVICE_NATIVE_VECTOR_WIDTH_INT:            break;
      case ocl_info.CL_DEVICE_NATIVE_VECTOR_WIDTH_LONG:           break;
      case ocl_info.CL_DEVICE_NATIVE_VECTOR_WIDTH_FLOAT:          break;
      case ocl_info.CL_DEVICE_NATIVE_VECTOR_WIDTH_DOUBLE:         throw new CLError(ocl_errors.CL_INVALID_VALUE);
      case ocl_info.CL_DEVICE_NATIVE_VECTOR_WIDTH_HALF:           throw new CLError(ocl_errors.CL_INVALID_VALUE);
      case ocl_info.CL_DEVICE_MAX_CLOCK_FREQUENCY:                break;
      case ocl_info.CL_DEVICE_ADDRESS_BITS:                       break;
      case ocl_info.CL_DEVICE_MAX_MEM_ALLOC_SIZE:                 break;
      case ocl_info.CL_DEVICE_IMAGE_SUPPORT:                      return true;
      case ocl_info.CL_DEVICE_MAX_READ_IMAGE_ARGS:                break;
      case ocl_info.CL_DEVICE_MAX_WRITE_IMAGE_ARGS:               break;
      case ocl_info.CL_DEVICE_IMAGE2D_MAX_WIDTH:                  break;
      case ocl_info.CL_DEVICE_IMAGE2D_MAX_HEIGHT:                 break;
      case ocl_info.CL_DEVICE_IMAGE3D_MAX_WIDTH:                  return 0;
      case ocl_info.CL_DEVICE_IMAGE3D_MAX_HEIGHT:                 return 0;
      case ocl_info.CL_DEVICE_IMAGE3D_MAX_DEPTH:                  return 0;
      case ocl_info.CL_DEVICE_MAX_SAMPLERS:                       break;
      case ocl_info.CL_DEVICE_MAX_PARAMETER_SIZE:                 break;
      case ocl_info.CL_DEVICE_MEM_BASE_ADDR_ALIGN:                break;
      //case ocl_info.CL_DEVICE_MIN_DATA_TYPE_ALIGN_SIZE:
      case ocl_info.CL_DEVICE_SINGLE_FP_CONFIG:                   break;
      case ocl_info.CL_FP_ROUND_TO_NEAREST:                       break;
      case ocl_info.CL_DEVICE_GLOBAL_MEM_CACHE_TYPE:              break;
      case ocl_info.CL_DEVICE_GLOBAL_MEM_CACHELINE_SIZE:          break;
      case ocl_info.CL_DEVICE_GLOBAL_MEM_CACHE_SIZE:              break;
      case ocl_info.CL_DEVICE_GLOBAL_MEM_SIZE:                    break;
      case ocl_info.CL_DEVICE_MAX_CONSTANT_BUFFER_SIZE:           break;
      case ocl_info.CL_DEVICE_MAX_CONSTANT_ARGS:                  break;
      case ocl_info.CL_DEVICE_LOCAL_MEM_TYPE:                     break;
      case ocl_info.CL_DEVICE_LOCAL_MEM_SIZE:                     break;
      case ocl_info.CL_DEVICE_ERROR_CORRECTION_SUPPORT:           break;
      case ocl_info.CL_DEVICE_HOST_UNIFIED_MEMORY:                break;
      case ocl_info.CL_DEVICE_PROFILING_TIMER_RESOLUTION:         break;
      case ocl_info.CL_DEVICE_ENDIAN_LITTLE:                      break;
      case ocl_info.CL_DEVICE_AVAILABLE:                          break;
      case ocl_info.CL_DEVICE_COMPILER_AVAILABLE:                 break;
      case ocl_info.CL_DEVICE_EXECUTION_CAPABILITIES:             return ocl_const.CL_EXEC_KERNEL;
      case ocl_info.CL_DEVICE_QUEUE_PROPERTIES:                   break;
      case ocl_info.CL_DEVICE_PLATFORM:                           break;
      case ocl_info.CL_DEVICE_NAME:                               break;
      case ocl_info.CL_DEVICE_VENDOR:                             break;
      case ocl_info.CL_DRIVER_VERSION:                            break;
      case ocl_info.CL_DEVICE_PROFILE:                            return "WEBCL_PROFILE";
      case ocl_info.CL_DEVICE_VERSION:                            return "WebCL 1.0";
      case ocl_info.CL_DEVICE_OPENCL_C_VERSION:                   return "WebCL C 1.0";
      case ocl_info.CL_DEVICE_EXTENSIONS:                         break;
      default:
        throw new CLError (ocl_errors.CL_INVALID_VALUE, "", "WebCLDevice.getInfo");
    }

    return this._wrapInternal (this._internal.getInfo (name));
  }
  catch (e)
  {
    try { ERROR(String(e)); }catch(e){}
    throw webclutils.convertCLException (e);
  }
};


Device.prototype.getSupportedExtensions = function ()
{
  TRACE (this, "getSupportedExtensions", arguments);
  if(!this._ensureValidObject ()) throw new CLInvalidated();

  try
  {
    // TODO!
    return [];
  }
  catch (e)
  {
    try { ERROR(String(e)); }catch(e){}
    throw webclutils.convertCLException (e);
  }
};


Device.prototype.enableExtension = function (extensionName)
{
  TRACE (this, "enableExtension", arguments);
  if(!this._ensureValidObject ()) throw new CLInvalidated();

  try
  {
    // TODO;
    return false;
  }
  catch (e)
  {
    try { ERROR(String(e)); }catch(e){}
    throw webclutils.convertCLException (e);
  }
};


Device.prototype._getRefCount = function ()
{
  // WebCLDevice doesn't support refcounting.
  return 0;
};



var NSGetFactory = XPCOMUtils.generateNSGetFactory ([Device]);


} catch(e) { ERROR ("webcldevice.js: "+e); throw e; }
